File: rtree/RangeSolver.java

public class RangeSolver {
    public static class Pos {
        public final int x;
        public final int y;

        public Pos(int x, int y) {
            this.x = x;
            this.y = y;
        }

        public String toString() {
            return "(" + x + "," + y + ")";
        }
    }
    
    public static class Zone {
        public static final Zone none() {
            return new Zone(null, null);
        }
        
        public final Pos nw;
        public final Pos se;

        public Zone(int w, int n, int e, int s) {
            this.nw = new Pos(w, n);
            this.se = new Pos(e, s);
        }

        public Zone(Pos nw, Pos se) {
            this.nw = nw;
            this.se = se;
        }

        public boolean contains(Pos pos) {
            if(this.nw == null || this.se == null) {
                return false;
            }
            
            if(pos.x < nw.x) return false;
            if(pos.x > se.x) return false;
            if(pos.y < nw.y) return false;
            if(pos.y > se.y) return false;
            return true;
        }

        public boolean overlaps(Zone zone) {
            if(this.nw == null || this.se == null) {
                return false;
            }
            
            if(zone.nw == null || zone.se == null) {
                return false;
            }
            
            if(this.nw.x > zone.se.x) return false;
            if(this.nw.y > zone.se.y) return false;
            if(this.se.x < zone.nw.x) return false;
            if(this.se.y < zone.nw.y) return false;
            return true;
        }

        public Zone extendTo(Pos pos) {
            Pos nw, se;
            
            if(this.nw == null) {
                nw = pos;
            } else {
                int x = Math.min(this.nw.x, pos.x);
                int y = Math.min(this.nw.y, pos.y);
                nw = new Pos(x, y);
            }
            
            if(this.se == null) {
                se = pos;
            } else {
                int x = Math.max(this.se.x, pos.x);
                int y = Math.max(this.se.y, pos.y);
                se = new Pos(x, y);
            }

            return new Zone(nw, se);
        }

        public Zone extendTo(Zone zone) {
            return this.extendTo(zone.nw).extendTo(zone.se);
        }

        public String toString() {
            return "[Zone " + nw + ", " + se + "]";
        }
    }

    public static class CellState {
        public final int state;

        public static final CellState gas = new CellState(1);
        public static final CellState liquid = new CellState(2);
        public static final CellState solid = new CellState(3);

        private CellState(int state) {
            this.state = state;
        }
    }
    
    public final Map<Pos, CellState> grid = new HashMap<>();
    public static class Cell {
        public final Pos pos;
        public CellState state;
    }

    public static class Hint {
        public int value;
        public List<Map<Pos, CellState>>
    }
    
    public static class Solver {
        public final Map<Pos, CellState> grid;
        public 
    }
}

File: rtree/RangeSolver.hs

module RangeSolver
(
) where

import qualified Zone
import Zone (Zone(..), Pos)
import qualified TRTree
import TRTree (Tree(..))

-- We have the following data structures:
-- 1) The grid. Each Pos has exactly one cell. Each cell has a Pos and a state -- liquid (uncertain) by default.
-- 2) The hints. Each hint has:
--   - a horz zone
--   - a vert zone
--   - a value
-- 3) The hints are stored as Clear or Dirty hints.
-- 4) Changing a cell dirties all the hints that govern that cell.

data CellState = Solid | Liquid | Gas
type Grid = Tree CellState
data Cell = Cell Pos CellState
data Hint = Hint Int (Tree Pos) (Tree Pos) -- value, horz_zone, vert_zone
data SolverState = SolverState Grid [Hint] [Hint] -- cells, clear, dirty

init_grid :: Int -> Int -> Grid

create_hint :: Int -> Pos -> Hint

init_hint :: Hint -> SolverState -> SolverState

solve :: SolverState -> Maybe SolverState
solve (SolverState grid [] [] resolved) = Nothing
solve (SolverState grid [] (hint:potential) resolved) = case find_hint_value grid hint of
                                                             Nothing -> solve $ SolverState grid [] potential (hint:resolved)
                                                             Just replacement_cell -> set_cell_in_state replacement_cell (SolverState grid [] (hint:potential) resolved)
solve (SolverState grid (hint:outdated) potential resolved) = case update_hint grid hint of
                                                                   Nothing -> solve $ SolverState grid outdated (hint:potential) resolved
                                                                   Just replacement_hint -> SolverState grid outdated (replacement_hint:potential) resolved

find_hint_value :: Grid -> Hint -> (Maybe Hint, Maybe Cell)
find_hint_value grid hint = find_hint_value_x grid (choose_first_liquids grid hint)

find_hint_value_x :: Grid -> [Hint] -> [Cell] -> Maybe Cell
find_hint_value_x _ _ [] = Nothing
find_hint_value_x grid hints (cell:cells)

-- solve algorithm:
-- 1) if there are no dirty hints, we are Done (succeeded or failed)
-- 2) pick any dirty hint (arbitrarily)
--     2a) reduce the hint's horz and vert sizes (if there are any solids in them)
--     2b) pick each first liquid cell in each of west/north/east/south (up to 4 of these)
--         2a1) if there are no liquid cells in it, then destroy this hint
--         2a2) for each of these liquids, try setting it to Solid and then to Gas
--             2a2a) when changing this cell, lookup every dirty and clear hint that governs this cell
--             2a2b) if any of these hints are contradicted, the Liquid state is disproven and we return the new value
--     2c) if a new value for one of these liquids was picked, return it (along with the set of hints that will now be dirty?)
--     2d) if none were found, clear the dirt on this hint
